pkg/kubelet/certificate/kubelet.go

```diff
@@ -70,9 +72,32 @@ func newGetTemplateFn(nodeName types.NodeName, getAddresses func() []v1.NodeAddr
 	}
 }
 
+// getNodeAddressesFromInformer returns the node addresses from the informer cache.
+// On external cloud providers nodes it waits for the external cloud provider to
+// remove the taint, since the node may start with some initial addresses that are changed
+// later by the cloud provider and we need to use only the definitive addresses for the
+// certificate request.
+func getNodeAddressesFromInformer(nodeName types.NodeName, nodeLister corelisters.NodeLister) []v1.NodeAddress {
+	node, err := nodeLister.Get(string(nodeName))
+	if err != nil {
+		klog.ErrorS(err, "fail to obtain node from local cache", "node", nodeName)
+		return nil
+	}
+	if len(node.Status.Addresses) == 0 {
+		return nil
+	}
+	// wait for the cloud provider to initialize the node addresses
+	for _, taint := range node.Spec.Taints {
+		if taint.Key == cloudproviderapi.TaintExternalCloudProvider {
+			return nil
+		}
+	}
+	return node.Status.Addresses
+}
+
 // NewKubeletServerCertificateManager creates a certificate manager for the kubelet when retrieving a server certificate
 // or returns an error.
-func NewKubeletServerCertificateManager(kubeClient clientset.Interface, kubeCfg *kubeletconfig.KubeletConfiguration, nodeName types.NodeName, getAddresses func() []v1.NodeAddress, certDirectory string) (certificate.Manager, error) {
```

Comment:
you shouldn't need to change this file, just change the function passed in to get node addresses, right?